### PR TITLE
Implement leader election algorithm.

### PR DESCRIPTION
**Pull request by vldm**
Closes: #30, https://github.com/stegos/stegos/issues/39

Elect new leader based on randhound random.
This pr contain a mock of validator stake, and group forming code.

### CODE REVIEW

File: node/src/lib.rs [hunk 1]
```diff
@@ -138,7 +139,7 @@ impl Node {
 pub struct EpochNotification {
     pub epoch: u64,
     pub leader: SecurePublicKey,
-    pub witnesses: Vec<SecurePublicKey>,
+    pub witnesses: Vec<ValidatorState>,
```
Comment (rtsisyk) [inline node/src/lib.rs]: I comment this part in #249. Please check.

File: node/src/lib.rs [hunk 2]
```diff
@@ -571,6 +677,41 @@ impl NodeService {
...
+        //TODO: replace by real calculation.
+        self.stakes.get(validator).cloned().unwrap_or(0)
+    }
+
+    /// Adds some stake to node full stake.
+    #[allow(dead_code)]
+    fn add_stake(&mut self, node: &SecurePublicKey, stake: u64) {
+        *self.stakes.entry(*node).or_insert(0) += stake;
+        //update current validator list
+        for validator in &mut self.witnesses {
```
Comment (rtsisyk) [inline node/src/lib.rs]: This eenfficient loop will go away after reworking NodeService.witnesses array according to comments in #249

File: node/src/lib.rs [hunk 3]
```diff
@@ -551,16 +601,72 @@ impl NodeService {
...
     fn handle_new_randomness(&mut self, h: Hash) -> Result<(), Error> {
-        self.randomness = h;
+        self.randomness.push(h);
         info!("New randomness obtained: {}", h);
-        // Start new round, if we are the leader
-        if self.is_leader() {
-            RandHound::start_round(&self.randhound)?;
+        if self.randomness.len() == self.election_round && self.is_leader() {
+            debug_assert!(self.election_round > 0);
+            let last_random = *self.randomness.get(self.randomness.len() - 1).unwrap();
```
Comment (rtsisyk) [inline node/src/lib.rs]: May be randomness.last().unwrap()?

File: node/src/lib.rs [hunk 4]
```diff
@@ -551,16 +601,72 @@ impl NodeService {
...
 
+    /// Handler for RandHound random source.
     fn handle_new_randomness(&mut self, h: Hash) -> Result<(), Error> {
-        self.randomness = h;
+        self.randomness.push(h);
         info!("New randomness obtained: {}", h);
-        // Start new round, if we are the leader
-        if self.is_leader() {
-            RandHound::start_round(&self.randhound)?;
+        if self.randomness.len() == self.election_round && self.is_leader() {
```
Comment (rtsisyk) [inline node/src/lib.rs]: I don't see there self.election_round is incremented. It seems to be constant right now. Why don't convert it to a constant?

File: node/src/lib.rs [hunk 5]
```diff
@@ -551,16 +601,72 @@ impl NodeService {
...
+                RandHound::start_round(&self.randhound)?;
+            }
         }
         Ok(())
     }
 
+    /// Handler for new epoch creation procedure.
+    /// This method called only on leader side.
+    /// Leader should create a KeyBlock based on last random provided by randhound.
+    fn handle_create_new_epoch(&self, random: Hash) -> KeyBlock {
```
Comment (rtsisyk) [inline node/src/lib.rs]: I use prefix "handle_" for all functions which handle incoming requests from event loop.

File: node/src/lib.rs [hunk 6]
```diff
@@ -571,6 +677,41 @@ impl NodeService {
             .retain(move |tx| tx.unbounded_send(balance).is_ok())
     }
 
+    /// Returns stake from database.
+    /// Called when witnesses list was changed.
+    fn validator_stake(&self, validator: &SecurePublicKey) -> u64 {
+        //TODO: replace by real calculation.
```
Comment (rtsisyk) [inline node/src/lib.rs]: I don't understand this TODO. `validator_stake()` just returns a calculated stake for specific validator. What else should be done in this function?

File: node/src/lib.rs [hunk 7]
```diff
@@ -551,16 +601,72 @@ impl NodeService {
...
+            info!("Created block: hash={}", Hash::digest(&key_block));
+
+            //
+            // Seal and register the block.
+            //
+
+            let block2 = key_block.clone();
+            self.chain
+                .register_key_block(block2)
+                .expect("mempool transaction are validated before");
```
Comment (rtsisyk) [inline node/src/lib.rs]: .except() has wrong comment.

**Comment by vldm** [comment]
@rtsisyk i rewrite stakes initialisation code, for 3 nodes it produce:

`2018-12-17 12:07:45 [stegos_node] DEBUG stegos_node:Stake holders = {SecurePKey(2dc146785191510447003812fb8be095e468d6b0e36742611ad6ecfb7cde5b9c63202131a5093085f96c4de5bf2ccf47898e1d59206eeb003abb4b592e2dfe6901): 15, SecurePKey(635313d8cf2b710ec7cf07121c46c72c565d42478aaa267db2960a466a2ac97ebf695ea2d50cff6d6b1f12ed64476841e21794d9da0f439604d44d78958dbd5900): 7, SecurePKey(73f7c2c212b28c580f3b30ea529e508575736455127775fd934fcc59205e9bd44ef9ffb42450591329d6d7090c71bf4ea16b73085f8e6173b40d26b45cf6c60e01): 3}
`
15, 7, 3

File: node/src/lib.rs [hunk 8]
```diff
@@ -551,16 +576,69 @@ impl NodeService {
...
+        let stakes: Vec<_> = witnesses
+            .iter()
+            .map(|key| self.validator_stake(key))
+            .collect();
+
+        let shrinked_random = election::shrink_random(random);
+        let leader = match election::elect_new_leader(&stakes, shrinked_random) {
+            Some(val) => witnesses[val],
+            None => {
+                error!("Cannot calculate new leader, fallback to first validator.");
```
Comment (rtsisyk) [inline node/src/lib.rs]: I think that we should panic in this case. I have no better idea how to handle this situation.

**Pull request merged**